extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -16,45 +16,176 @@
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.EncoderFactory;
 import org.apache.avro.util.Utf8;
 
+import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+import static com.hazelcast.jet.impl.util.ExceptionUtil.rethrow;
+import static org.apache.avro.SchemaNormalization.toParsingForm;
 
 /**
- * Hazelcast serializer hooks for the classes in the {@code
- * com.hazelcast.jet.avrp} package. This is not a public-facing API.
+ * Hazelcast serializer hooks for the classes in the {@code com.hazelcast.jet.avro} package.
  */
-class AvroSerializerHooks {
+public final class AvroSerializerHooks {
+    private static final int SCHEMA_ID_SIZE = 8;
 
-    public static final class Utf8Hook implements SerializerHook<Utf8> {
+    private AvroSerializerHooks() { }
+
+    public static <T> byte[] serialize(DatumWriter<T> datumWriter, T data) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+            serialize(out, datumWriter, data);
+            return out.toByteArray();
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static <T> void serialize(ByteArrayOutputStream out, DatumWriter<T> datumWriter, T data) throws IOException {
+        BinaryEncoder encoder = EncoderFactory.get().directBinaryEncoder(out, null);
+        datumWriter.write(data, encoder);
+        encoder.flush();
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data) {
+        return deserialize(datumReader, data, 0, data.length);
+    }
+
+    public static <T> T deserialize(DatumReader<T> datumReader, byte[] data, int offset, int length) {
+        try {
+            BinaryDecoder decoder = DecoderFactory.get().binaryDecoder(data, offset, length, null);
+            return datumReader.read(null, decoder);
+        } catch (IOException e) {
+            throw rethrow(e);
+        }
+    }
+
+    public static class GenericRecordHook implements SerializerHook<GenericRecord>, HazelcastInstanceAware {
+        private AvroSchemaStorage schemas;
+
+        @Override
+        public Class<GenericRecord> getSerializationType() {
+            return GenericRecord.class;
+        }
+
+        @Override
+        public Serializer createSerializer() {
+            return new ByteArraySerializer<GenericRecord>() {
+                @Override
+                public int getTypeId() {
+                    return SerializerHookConstants.AVRO_GENERIC_RECORD;
+                }
+
+                @Override
+                public byte[] write(GenericRecord record) {
+                    if (record instanceof LazyImmutableRecord) {
+                        return ((LazyImmutableRecord) record).serializedRecord;
+                    }
+                    try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
+                        long schemaId = schemas.put(record.getSchema());
```

Comment:
is this PR mostly a performance optimisation of Avro GenericRecord serde with the baseline implementation being just sending each record along with schema?